src/cargo/ops/cargo_package/vcs.rs

```diff
@@ -111,6 +113,50 @@ pub fn check_repo_state(
     return Ok(Some(VcsInfo { git, path_in_vcs }));
 }
 
+/// Warns if any symlinks were checked out as plain text files.
+///
+/// Git config [`core.symlinks`] defaults to true when unset.
+/// In git-for-windows (and git as well),
+/// the config should be set to false explicitly when the repo was created,
+/// if symlink support wasn't detected [^1].
+///
+/// We assume the config was always set at creation time and never changed.
+/// So, if it is true, we don't bother users with any warning.
+///
+/// [^1]: <https://github.com/git-for-windows/git/blob/f1241afcc7956918d5da33ef74abd9cbba369247/setup.c#L2394-L2403>
+///
+/// [`core.symlinks`]: https://git-scm.com/docs/git-config#Documentation/git-config.txt-coresymlinks
+fn warn_symlink_checked_out_as_plain_text_file(
+    gctx: &GlobalContext,
+    src_files: &[PathEntry],
+    repo: &git2::Repository,
+) -> CargoResult<()> {
+    let core_symlinks_disabled = match repo.config().and_then(|c| c.get_bool("core.symlinks")) {
+        Ok(enabled) => !enabled,
+        Err(e) => {
+            debug!(
+                "no git config `core.symlinks` set for repo at `{}`: {e}",
+                repo.path().display(),
+            );
+            return Ok(());
+        }
+    };
```

Comment:
I'm not sure if this is an improvement but thought I'd share it in case you had thoughts:
```rust
if !repo.config().and_then(|c| c.get_bool("core.symlinks")).unwrap_or(true) {
    return Ok(());
}